sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/json/JsonFileFormat.scala

```diff
@@ -114,7 +114,16 @@ class JsonFileFormat extends TextBasedFileFormat with DataSourceRegister {
     }
 
     (file: PartitionedFile) => {
-      val parser = new JacksonParser(actualSchema, parsedOptions)
+      // SPARK-21610: when the `requiredSchema` only contains `_corrupt_record`,
```

Comment:
does this bug apply for csv too?

Reply:
> We need to let users know that _corrupted_record is a derived column from other columns and cannot be selected alone in a query.

@viirya I created issue https://issues.apache.org/jira/browse/SPARK-21610 and need to select field "_corrupt_record" alone. This is possible with spark 2.2 (if a dataframe is created from a RDD) and it would be great to keep this behaviour in future versions of Spark.

My use case is the following one: a spark job reads JSON with an input schema, and will: 
- save records that match the input schema in parquet format
- save "corrupt records" (invalid JSON or records that do match the input schema) to text files, in a separate folder.

Basically, I want : 
- a folder with clean data in parquet format
- another folder with "corrupt records". I can then analyze corrupt records and for instance tell partners that they are sending invalid data. This enables a clean data pipeline that separates valid records from corrupt records

To get valid and corrupt records, I write : 
```
      val validRecords = df.filter(col("_corrupt_record").isNull)
        .drop("_corrupt_record")

      val corruptRecords = df.filter(col("_corrupt_record").isNotNull)
        .select("_corrupt_record")
```

Reply:
Your usage scenario makes sense to me. The contents of `_corrupt_record ` depends on the fields our parser passed. The workaround is you can save your output to the cache or a physical table. 

```Scala
    val df = dfFromFile.cache()
    df.filter($"_corrupt_record".isNull).drop("_corrupt_record").show()
    df.filter($"_corrupt_record".isNotNull).select("_corrupt_record").show()
```

My current suggestion is to capture the empty actual schema and issue an error with a reasonable workaround message. Users can at least know what happened and how to fix the issue.

Reply:
@dm-tran I think the current change can support your use cases. When only `_corrupt_record` is selected, it is as the same as the effect of selecting all columns, i.e. a record is recognized as corrupt if any column is in invalid format.

However, it seems be countering the designed behavior that lets `_corrupt_record` depend on the selected columns, as we discussed in previous comments.

I think @gatorsmile's suggestion should be good.

Reply:
@gatorsmile @viirya I also think that @gatorsmile's suggestion looks good. Thanks for your replies!